course_UvA-DL/transformers-and-MH-attention/Transformers_and_MHAttention.py

```diff
@@ -0,0 +1,1479 @@
+# Next, we want to run the pre-trained ResNet model on the images, and extract the features before the classification layer.
+# These are the most high-level features, and should sufficiently describe the images.
+# CIFAR100 has some similarity to ImageNet, and thus we are not retraining the ResNet model in any form.
+# However, if you would want to get the best performance and have a very large dataset, it would be better to add the ResNet to the computation graph during training and finetune its parameters as well.
+# As we don't have a large enough dataset and want to train our model efficiently, we will extract the features beforehand.
+# Let's load and prepare the model below.
+
+# %%
+os.environ["TORCH_HOME"] = CHECKPOINT_PATH
+pretrained_model = torchvision.models.resnet34(pretrained=True)
+# Remove classification layer
+# In some models, it is called "fc", others have "classifier"
+# Setting both to an empty sequential represents an identity map of the final features.
+pretrained_model.fc = nn.Sequential()
+pretrained_model.classifier = nn.Sequential()
+# To GPU
+pretrained_model = pretrained_model.to(device)
+
+# Only eval, no gradient required
+pretrained_model.eval()
+for p in pretrained_model.parameters():
+    p.requires_grad = False
+
+# %% [markdown]
+# We will now write a extraction function for the features below.
+# This cell requires access to a GPU, as the model is rather deep and the images relatively large.
+# The GPUs on GoogleColab are sufficient, but running this cell can take 2-3 minutes.
+# Once it is run, the features are exported on disk so they don't have to be recalculated every time you run the notebook.
+# However, this requires >150MB free disk space.
+# So it is recommended to run this only on a local computer if you have enough free disk and a GPU (GoogleColab is fine for this).
+# If you do not have a GPU, you can download the features from the [GoogleDrive folder](https://drive.google.com/drive/folders/1DF7POc6j03pRiWQPWSl5QJX5iY-xK0sV?usp=sharing).
+
+
+# %%
+@torch.no_grad()
+def extract_features(dataset, save_file):
+    if not os.path.isfile(save_file):
+        data_loader = data.DataLoader(dataset, batch_size=128, shuffle=False, drop_last=False, num_workers=4)
+        extracted_features = []
+        for imgs, _ in tqdm(data_loader):
+            imgs = imgs.to(device)
+            feats = pretrained_model(imgs)
+            extracted_features.append(feats)
+        extracted_features = torch.cat(extracted_features, dim=0)
+        extracted_features = extracted_features.detach().cpu()
+        torch.save(extracted_features, save_file)
+    else:
+        extracted_features = torch.load(save_file)
+    return extracted_features
+
+
+train_feat_file = os.path.join(CHECKPOINT_PATH, "train_set_features.tar")
+train_set_feats = extract_features(train_set, train_feat_file)
+
+test_feat_file = os.path.join(CHECKPOINT_PATH, "test_set_features.tar")
+test_feats = extract_features(test_set, test_feat_file)
+
```

Comment:
The tests currently fail because we have one computationally expensive cell (Cell 27). This cell applies a ImageNet pre-trained ResNet on the CIFAR dataset which is upscaled to 224x224 pixels. This is no problem on GoogleColab, but the tests here run out of time or memory. Any way we can make the tests work?

Reply:
ok, trying to set ` --shm-size=32g`